src/dcc.c

```diff
@@ -2239,24 +2239,30 @@ struct dcc_table DCC_IDENTWAIT = {
 
 void dcc_ident(int idx, char *buf, int len)
 {
-  char response[512], uid[512], buf1[UHOSTLEN + 21];
+  char response[512], buf1[512], *c;
+  size_t l;
   int i;
 
-  *response = *uid = '\0';
-  sscanf(buf, "%*[^:]:%[^:]:%*[^:]:%[^\n]\n", response, uid);
+  *response = *buf1 = '\0';
+  sscanf(buf, "%*[^:]:%[^:]:%*[^:]:%[^\n]\n", response, buf1);
   rmspace(response);
   if (response[0] != 'U') {
     dcc[idx].timeval = now;
     return;
   }
-  rmspace(uid);
-  uid[20] = 0;                  /* 20 character ident max */
-  for (i = 0; i < dcc_total; i++)
-    if ((dcc[i].type == &DCC_IDENTWAIT) &&
-        (dcc[i].sock == dcc[idx].u.ident_sock)) {
-      snprintf(buf1, sizeof buf1, "%s@%s", uid, dcc[idx].host);
-      dcc_telnet_got_ident(i, buf1);
-    }
+  rmspace(buf1);
+  buf1[(sizeof buf1) - 1] = 0;
+  if (!strchr(buf1, '@')) {
+    c = buf1 + strlen(buf1);
+    *c++ = '@';
```

Comment:
if buf1 is a full 512 characters, is this going to overflow? should buf1 be 513?

Reply:
I modified sscanf to read up to 511 chars. lets keep buffer sizes to powers of 2.